packages/babel-plugin-proposal-pipeline-operator/src/smartVisitor.js

```diff
@@ -1,38 +1,48 @@
 import { types as t } from "@babel/core";
-import minimalVisitor from "./minimalVisitor";
 
 const updateTopicReferenceVisitor = {
   PipelinePrimaryTopicReference(path) {
     path.replaceWith(this.topicId);
   },
-  AwaitExpression(path) {
-    throw path.buildCodeFrameError(
-      "await is not supported inside pipeline expressions yet",
-    );
-  },
-  YieldExpression(path) {
-    throw path.buildCodeFrameError(
-      "yield is not supported inside pipeline expressions yet",
-    );
-  },
   PipelineTopicExpression(path) {
     path.skip();
   },
 };
 
 const smartVisitor = {
-  ...minimalVisitor,
-  PipelineTopicExpression(path) {
-    const topicId = path.scope.generateUidIdentifier("topic");
+  BinaryExpression: {
+    exit(path) {
+      const { scope } = path;
+      const { node } = path;
+      const { operator, left } = node;
+      let { right } = node;
+      if (operator !== "|>") return;
+
+      if (t.isIdentifier(right, { name: "eval" })) {
+        right = t.sequenceExpression([t.numericLiteral(0), right]);
+      }
+
+      const placeholder = scope.generateUidIdentifierBasedOnNode(left);
+      scope.push({ id: placeholder });
 
-    path.traverse(updateTopicReferenceVisitor, { topicId });
+      let call;
+      if (t.isPipelineTopicExpression(right)) {
+        path
+          .get("right")
+          .traverse(updateTopicReferenceVisitor, { topicId: placeholder });
 
-    const arrowFunctionExpression = t.arrowFunctionExpression(
-      [topicId],
-      path.node.expression,
-    );
+        call = right.body ? right.body : right.expression;
+      } else {
+        call = t.callExpression(right, [t.cloneNode(placeholder)]);
```

Comment:
Here you can use `right.callee`, so the separate `PipelineBareFunction` (and thus `exit(`) can be removed.

Reply:
Good idea. Will try that.